Spigot-Server-Patches/0390-Add-Velocity-IP-Forwarding-Support.patch

```diff
@@ -0,0 +1,308 @@
+@@ -187,6 +189,14 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+             this.g = LoginListener.EnumProtocolState.KEY;
+             this.networkManager.sendPacket(new PacketLoginOutEncryptionBegin("", this.server.E().getPublic(), this.e));
+         } else {
++            // Paper start - Velocity support
++            if (com.destroystokyo.paper.PaperConfig.velocitySupport) {
++                this.velocityLoginMessageId = java.util.concurrent.ThreadLocalRandom.current().nextInt();
++                PacketLoginOutCustomPayload packet = new PacketLoginOutCustomPayload(this.velocityLoginMessageId, com.destroystokyo.paper.proxy.VelocityProxy.PLAYER_INFO_CHANNEL, new PacketDataSerializer(io.netty.buffer.Unpooled.EMPTY_BUFFER));
++                this.networkManager.sendPacket(packet);
++                return;
++            }
++            // Paper end
+             // Spigot start
+             // Paper start - Cache authenticator threads
+             authenticatorPool.execute(new Runnable() {
+@@ -278,6 +288,12 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     public class LoginHandler {
+ 
+         public void fireEvents() throws Exception {
++                            // Paper start - Velocity support
++                            if (LoginListener.this.velocityLoginMessageId == -1 && com.destroystokyo.paper.PaperConfig.velocitySupport) {
++                                disconnect("This server requires you to connect with Velocity.");
++                                return;
++                            }
++                            // Paper end
+                             String playerName = i.getName();
+                             java.net.InetAddress address = ((java.net.InetSocketAddress) networkManager.getSocketAddress()).getAddress();
+                             java.util.UUID uniqueId = i.getId();
+@@ -325,6 +341,35 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     // Spigot end
+ 
+     public void a(PacketLoginInCustomPayload packetloginincustompayload) {
++        // Paper start - Velocity support
++        if (com.destroystokyo.paper.PaperConfig.velocitySupport && packetloginincustompayload.getId() == this.velocityLoginMessageId) {
++            PacketDataSerializer buf = packetloginincustompayload.getBuf();
++            if (buf == null) {
++                this.disconnect("This server requires you to connect with Velocity.");
++                return;
++            }
++
++            if (!com.destroystokyo.paper.proxy.VelocityProxy.checkSignature(buf)) {
++                this.disconnect("Unable to verify player details");
++                return;
++            }
++
++            this.networkManager.l = new java.net.InetSocketAddress(com.destroystokyo.paper.proxy.VelocityProxy.readAddress(buf), ((java.net.InetSocketAddress) this.networkManager.getSocketAddress()).getPort());
```

Comment:
Add an obfuscation helper as a setter right next to the field declaration please